src/python/pants/backend/python/subsystems/python_tool_base.py

```diff
@@ -221,6 +235,13 @@ def to_pex_request(
             sources=sources,
         )
 
+    @classmethod
+    def rules(cls: Any) -> Iterable[Any]:
+        yield from super().rules()
+        from pants.backend.python.util_rules import lockfile
```

Comment:
What about adding a protocol instead to break the import cycle:
```python
from Typing import Protocol

class PythonToolRequirementsBase(Protocol):
    """Break the import cycle."""
    options_scope: str
    lockfile_rules_type: LockfileRules
```

and then put `LockfileRules` in another file like `backend/python/target_types.py` or even `backend/python/subsystem/lockfile_rules.py` that both of these files can import from. Even once moved, I think all the subsystems should continue to import `LockfileRules` from `python_tool_base`, by convention.

The Protocol and `LockfileRules` could be defined in the same file (separate from `util_rules/lockfile.py` and `subsystem/python_tool_base.py`).

Reply:
I considered a protocol, especially since the helper rules don't really use many attributes of the `PythonToolRequirementsBase`. The problem comes when they forward to `GeneratePythonLockfile.from_tool`, which references many other attributes of `PythonToolRequirementsBase` (some of which are computed and reference even more). Some of these are exclusively lockfile related (`lockfile`, `uses_custom_lockfile`) which could be pulled out or naturally fit here. But others were more general, like `interpreter_constraints` or `all_requirements`. You end up depending on the entirety of `PythonToolRequirementsBase`.

Thinking along that line leads to the idea that maybe `PythonToolRequirementsBase` is already the Interface. It has many helpers to work with a lockfile already on it. Some of these do more than just option handling: `pex_requirements` looks similar to a mirror version of `GeneratePythonLockfile.from_tool`. So maybe it makes sense to pull `PythonToolRequirementsBase` and the helper rules out into one file, where they can mutually reference each other. 

Thinking about it more, though, if we moved `GeneratePythonLockfile.from_tool` onto `PythonToolRequirementsBase`, we could then have a protocol that was just
```python
class PythonToolRequirementsBase(Protocol):
    """Break the import cycle."""
    options_scope: str
    lockfile_rules_type: LockfileRules
    def to_generate_python_lockfile(...) -> GeneratePythonLockfile: ...
```
which actually seems pretty good. Thanks for encouraging me to have a second look at this. I hope this can be done without a breaking change to the plugins API, it seems to be used by every Python tool. I'll try to see if it works out.

Along the line of defining the Protocol and `LockfileRules` separately, if we split `PythonToolBase` and `PythonToolRequirementsBase`; and registered the rules in `PythonToolBase`; we could also linearise this. That's how it worked on my first pass at this, the concrete tool depended on the helpers and `PythonToolRequirementsBase`; and the helpers depended on `PythonToolRequirementsBase`

Reply:
Huh. Nice.

And `GeneratePythonLockfile.from_tool` is the only thing in backend.python.goal.lockfile that currently uses `PythonToolRequirementsBase`, so moving that onto `PythonToolRequirementsBase.xyz` will not create a circular import. That sounds much cleaner.

For the method name, I think you're inverting "from" in `from_tool` to get `to_generate_python_lockfile`, but that sounds odd to me ("to generate" sounds like using the infinitive form of the verb). So what about `to_generate_python_lockfile_request` or `generate_python_lockfile_request` or `to_lockfile_request` or `to_lockfile_generation_request`?